code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java

```diff
@@ -381,9 +386,104 @@ public long getServiceNetworkId(Service service) {
         return networkId;
     }
 
-    protected String getServiceInstanceName(Service service, int order) {
+    protected List<? extends String> getServiceInstancesNames(Service service, int requestedCount, boolean toAdd) {
+        List<Integer> usedIds = getServiceInstanceUsedOrderIds(service);
+        if (toAdd) {
+            Collections.sort(usedIds);
+        } else {
+            Collections.sort(usedIds, Collections.reverseOrder());
+        }
+        List<String> serviceNames = new ArrayList<>();
+        if (toAdd) {
+            int finalCount = 0;
+            // in situation when service with scale=3 has one container missing (it got destroyed outside of the
+            // service)
+            // and container names don't reflect the order, we should pick the least available number that is <=order
+            for (int i = 1; i < requestedCount + 1; i++) {
+                if (finalCount < requestedCount) {
+                    if (!usedIds.contains(i)) {
+                        serviceNames.add(generateServiceInstanceName(service, i));
+                        finalCount++;
+                        usedIds.add(i);
+                    }
+                }
+            }
+            // only after we got all "gap" names, get the rest
+            Collections.sort(usedIds);
```

Comment:
finalCount reflects number of instances we return; not just the number of gaps filled. And we do have to proceed in case finalCount < requested count. Think about situation like that:

Updated scale=5
Existing instances: env_svc_1, randomName1, env_svc3

So first we fill up the gap by creating the instance env_svc2. And then we create env_svc4 instance - after line 412